File: src/main/webapp/scripts/models/asset.js
function Asset()
{
	this.mObjAppUser				=	undefined;
	this.mObjCategory				=	undefined;
	this.mbChanged					=	undefined;
	this.mdDepreciatedValue			=	undefined;
	this.msDescription				=	undefined;
	this.mdtExpiration				=	undefined;
	this.miId						=	undefined;
	this.mObjImage					=	undefined;
	this.msItemNumber				=	undefined;
	this.mArrObjObservers			=	[];
	this.mdtPosted					=	undefined;
	this.mdRetailValue				=	undefined;
	this.mObjStatus					=	undefined;
	this.msTitle					=	undefined;
}

Asset.prototype.addObserver			=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers	=	[];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

Asset.copyFromJSON					=	function(objJSON, objAsset)
{
	if(objAsset == undefined)
		objAsset = new Asset();
	
	for(prop in objJSON)
		switch(prop)
			{
			case 'category':
				Category.copyFromJSON(eval("objJSON." + prop), objAsset.getCategory());
				break;
			case 'image':
				AssetImage.copyFromJSON(eval("objJSON." + prop), objAsset.getImage());
				break;
			case 'poster':
				AppUser.copyFromJSON(eval("objJSON." + prop), objAsset.getPoster());
				break;
			case 'status':
				AssetStatus.copyFromJSON(eval("objJSON." + prop), objAsset.getStatus());
				break;
			default:
				objAsset["set" + prop.substring(0, 1).toUpperCase() + prop.substring(1)].call(objAsset, eval("objJSON." + prop));
			}
	
	return(objAsset);
};

Asset.prototype.getCategory			=	function()
{
	if(this.mObjCategory == undefined)
		{
		this.mObjCategory = new Category();
		
		this.mObjCategory.addObserver((function(objThis)
		{
			return(function(objArgument)
			{
				objThis.notifyObservers();
			});
		})(this));
		}
	
	return(this.mObjCategory);
};

Asset.prototype.getDepreciatedValue	=	function()
{
	if(this.mdDepreciatedValue == undefined)
		this.mdDepreciatedValue = 0.00;
	
	return(this.mdDepreciatedValue);
};

Asset.prototype.getDescription		=	function()
{
	if(this.msDescription == undefined)
		this.msDescription = "";
	
	return(this.msDescription);
};

Asset.prototype.getExpirationDate	=	function()
{
	return(this.mdtExpiration);
};

Asset.prototype.getId				=	function()
{
	if(this.miId == undefined)
		this.miId = 0;
	
	return(this.miId);
};

Asset.prototype.getImage			=	function()
{
	if(this.mObjImage == undefined)
		{
		this.mObjImage = new AssetImage();
		
		this.mObjImage.addObserver((function(objThis)
			{
				return(function(objArgument)
				{
					objThis.notifyObservers();
				});
			})(this));
		}
	
	return(this.mObjImage);
};

Asset.prototype.getItemNumber		=	function()
{
	if(this.msItemNumber ==	undefined)
		this.msItemNumber = "";
	
	return(this.msItemNumber);
};

Asset.prototype.getPostedDate		=	function()
{
	return(this.mdtPosted);
};

Asset.prototype.getPoster				=	function()
{
	if(this.mObjAppUser == undefined)
		{
		this.mObjAppUser = new AppUser();
		
		this.mObjAppUser.addObserver((function(objThis)
		{
			return(function(objArgument)
			{
				objThis.notifyObservers();
			});
		})(this));
		}
	
	return(this.mObjAppUser);
};

Asset.prototype.getRetailValue		=	function()
{
	if(this.mdRetailValue == undefined)
		this.mdRetailValue = 0.00;
	
	return(this.mdRetailValue);
};

Asset.prototype.getStatus			=	function()
{
	if(this.mObjStatus == undefined)
		{
		this.mObjStatus = new AssetStatus();
		
		this.mObjStatus.addObserver((function(objThis)
		{
			return(function(objArgument)
			{
				objThis.notifyObservers();
			});
		})(this));
		}
	
	return(this.mObjStatus);
};

Asset.prototype.getTitle			=	function()
{
	if(this.msTitle == undefined)
		this.msTitle = "";
	
	return(this.msTitle);
};

Asset.prototype.hasChanged				=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged = false;
	
	return(this.mbChanged);
};

Asset.prototype.notifyObservers			=	function(objArgument)
{
	this.mbChanged = true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged = false;
};

Asset.prototype.packageToJSON			=	function()
{
	return(	{	id:					this.getId()
			,	poster:				this.getPoster().packageToJSON()
			,	itemNumber:			this.getItemNumber()
			,	title:				this.getTitle()
			,	description:		this.getDescription()
			,	category:			this.getCategory().packageToJSON()
			,	status:				this.getStatus().packageToJSON()
			,	depreciatedValue:	this.getDepreciatedValue()
			,	retailValue:		this.getRetailValue()
			,	postedDate:			this.getPostedDate()
			,	expirationDate:		this.getExpirationDate()
			,	image:				this.getImage().packageToJSON()
			});
};

Asset.prototype.setCategory				=	function(objCategory)
{
	if(objCategory == undefined)
		this.mObjCategory 				=	new Category();
	else
		this.mObjCategory 				=	objCategory;
	
	this.mObjCategory.addObserver((function(objThis)
	{
		return(function(objArgument)
		{
			objThis.notifyObservers();
		});
	})(this));
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setDepreciatedValue		=	function(dDepreciatedValue)
{
	if(dDepreciatedValue == undefined)
		this.mdDepreciatedValue 		=	0.00;
	else
		this.mdDepreciatedValue			=	dDepreciatedValue;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setDescription			=	function(sDescription)
{
	if(sDescription == undefined)
		this.msDescription 				=	"";
	else
		this.msDescription				=	sDescription;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setExpirationDate		=	function(dtExpiration)
{
	this.mdtExpiration					=	dtExpiration;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setId					=	function(iId)
{
	if(iId == undefined)
		this.miId						=	0;
	else
		this.miId						=	iId;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setImage				=	function(objImage)
{
	if(objImage == undefined)
		this.mObjImage					=	new AssetImage();
	else
		this.mObjImage					=	objImage;
	
	this.mObjImage.addObserver((function(objThis)
	{
		return(function(objArgument)
		{
			objThis.notifyObservers();
		});
	})(this));
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setItemNumber			=	function(sItemNumber)
{
	if(sItemNumber ==	undefined)
		this.msItemNumber 				=	"";
	else
		this.msItemNumber				=	sItemNumber;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setPostedDate			=	function(dtPosted)
{
	this.mdtPosted						=	dtPosted;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setPoster				=	function(objAppUser)
{
	if(objAppUser == undefined)
		this.mObjAppUser				=	objAppUser;
	else
		this.mObjAppUser				=	objAppUser;
	
	this.mObjAppUser.addObserver((function(objThis)
	{
		return(function(objArgument)
		{
			objThis.notifyObservers();
		});
	})(this));
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setRetailValue			=	function(dRetailValue)
{
	if(dRetailValue == undefined)
		this.mdRetailValue 				=	0.00;
	else
		this.mdRetailValue				=	dRetailValue;
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setStatus				=	function(objStatus)
{
	if(objStatus == undefined)
		this.mObjStatus 				=	new AssetStatus();
	else
		this.mObjStatus					=	objStatus;
	
	this.mObjStatus.addObserver((function(objThis)
	{
		return(function(objArgument)
		{
			objThis.notifyObservers();
		});
	})(this));
	
	this.notifyObservers();
	
	return(this);
};

Asset.prototype.setTitle				=	function(sTitle)
{
	if(sTitle == undefined)
		this.msTitle 					=	"";
	else
		this.msTitle					=	sTitle;
	
	this.notifyObservers();
	
	return(this);
};

File: src/main/java/net/draconia/askaround/domain/Asset.java
package net.draconia.askaround.domain;

import java.io.Serializable;

import java.math.BigDecimal;

import java.util.Date;

import net.draconia.askaround.util.CurrencySerializer;
import net.draconia.askaround.util.DateSerializer;

import org.codehaus.jackson.map.annotate.JsonSerialize;

public class Asset implements Serializable
{
	private static final long serialVersionUID = -4432542610644713603L;
	
	private AppUser mObjPoster;
	private AssetImage mObjImage;
	private AssetStatus mObjStatus;
	private BigDecimal mdDepreciatedValue, mdRetailValue;
	private Category mObjCategory;
	private Date mDtExpiration, mDtPosted;
	private Integer miId;
	private String msDescription, msItemNumber, msTitle;
	
	public Asset()
	{ }
	
	public Category getCategory()
	{
		if(mObjCategory == null)
			mObjCategory = new Category();
		
		return(mObjCategory);
	}
	
	@JsonSerialize(using=CurrencySerializer.class)
	public BigDecimal getDepreciatedValue()
	{
		if(mdDepreciatedValue == null)
			mdDepreciatedValue = new BigDecimal(0);
		
		return(mdDepreciatedValue);
	}
	
	public String getDescription()
	{
		if(msDescription == null)
			msDescription = "";
		
		return(msDescription);
	}
	
	/**
	 * getExpirationDate - Returns the expiration date for this Asset
	 * @return expiration date in the Asset record
	 */
	// The annotation permits the Jackson library to encode the date into the JSON
	@JsonSerialize(using = DateSerializer.class)
	public Date getExpirationDate()
	{
		if(mDtExpiration == null)
			mDtExpiration = new Date();
		
		return(mDtExpiration);
	}
	
	public int getId()
	{
		if(miId == null)
			miId = 0;
		
		return(miId);
	}
	
	public AssetImage getImage()
	{
		if(mObjImage == null)
			mObjImage = new AssetImage();
		
		return(mObjImage);
	}
	
	public String getItemNumber()
	{
		if(msItemNumber == null)
			msItemNumber = "";
		
		return(msItemNumber);
	}
	
	/**
	 * getPostedDate - Returns the expiration date for this Asset
	 * @return posted date in the Asset record
	 */
	// The annotation permits the Jackson library to encode the date into the JSON
	@JsonSerialize(using = DateSerializer.class)
	public Date getPostedDate()
	{
		if(mDtPosted == null)
			mDtPosted = new Date();
		
		return(mDtPosted);
	}
	
	public AppUser getPoster()
	{
		if(mObjPoster == null)
			mObjPoster = new AppUser();
		
		return(mObjPoster);
	}
	
	@JsonSerialize(using=CurrencySerializer.class)
	public BigDecimal getRetailValue()
	{
		if(mdRetailValue == null)
			mdRetailValue = new BigDecimal(0);
		
		return(mdRetailValue);
	}
	
	public AssetStatus getStatus()
	{
		if(mObjStatus == null)
			mObjStatus = new AssetStatus();
		
		return(mObjStatus);
	}
	
	public String getTitle()
	{
		if(msTitle == null)
			msTitle = "";
		
		return(msTitle);
	}
	
	public void setCategory(final Category objCategory)
	{
		if(objCategory == null)
			mObjCategory = new Category();
		else
			mObjCategory = objCategory;
	}
	
	public void setDepreciatedValue(final BigDecimal dDepreciatedValue)
	{
		if(dDepreciatedValue == null)
			mdDepreciatedValue = new BigDecimal(0);
		else
			mdDepreciatedValue = dDepreciatedValue;
	}
	
	public void setDescription(final String sDescription)
	{
		if(sDescription == null)
			msDescription = "";
		else
			msDescription = sDescription;
	}
	
	public void setExpirationDate(final Date dtExpiration)
	{
		if(dtExpiration == null)
			mDtExpiration = new Date();
		else
			mDtExpiration = dtExpiration;
	}
	
	public void setId(final Integer iId)
	{
		if(iId == null)
			miId = 0;
		else
			miId = iId;
	}
	
	public void setImage(final AssetImage objImage)
	{
		if(objImage == null)
			mObjImage = new AssetImage();
		else
			mObjImage = objImage;
	}
	
	public void setItemNumber(final String sItemNumber)
	{
		if(sItemNumber == null)
			msItemNumber = "";
		else
			msItemNumber = sItemNumber;
	}
	
	public void setPostedDate(final Date dtPosted)
	{
		if(dtPosted == null)
			mDtPosted = new Date();
		else
			mDtPosted = dtPosted;
	}
	
	public void setPoster(final AppUser objPoster)
	{
		if(objPoster == null)
			mObjPoster = new AppUser();
		else
			mObjPoster = objPoster;
	}
	
	public void setRetailValue(final BigDecimal dRetailValue)
	{
		if(dRetailValue == null)
			mdRetailValue = new BigDecimal(0);
		else
			mdRetailValue = dRetailValue;
	}
	
	public void setStatus(final AssetStatus objStatus)
	{
		if(objStatus == null)
			mObjStatus = new AssetStatus();
		else
			mObjStatus = objStatus;
	}
	
	public void setTitle(final String sTitle)
	{
		if(sTitle == null)
			msTitle = "";
		else
			msTitle = sTitle;
	}
}